### DISCUSSION TITLE
Which ESP32?

### DISCUSSION THREAD

**Discussion by IanBUK** [Bodmer/TFT_eSPI#1418] [General]
I’ve kinda given up trying to get my Arduino Nano 33 BLE working with TFT_eSPI and my BuyDisplay 7” SSD1963 display. So…time to buy an ESP32 based controller.

ideally I’d prefer to go for a 16bit parallel connection - but wouldn’t that require 21 gpio pins plus the 2 for the i2c connection to the touchscreen? So, 23 input/output gpio pins.

I’d also prefer a board with built in lipo charger and BLE. 

Any recommendations please?

Thanks,

Ian Brown.

**Comment by Bodmer** [comment]
Hi,
TFT_eSPI only supports 8 bit parallel on the ESP32. 16 bit could be implmented but you would need to check the pin count needed which is a bit marginal in my view. This is a good resource:
https://randomnerdtutorials.com/esp32-pinout-reference-gpios/

Not that some pins are input only and this has been a "Gotcha" for some users"

There are a few ESP32 boards about with Lipo chargers, "ESP32 LiPo" should produce some options on Google such as the TTGO T-Energy.

If cost is not a barrier then Adafruit and SparkFun have products. I like SparkFun stuff myself and have a few of their products.

**Reply by IanBUK** [reply]
Evening...

Yes, I decided that I was spending so much time trying to make it work that, well the cost of just buying an Esp32 compared with the cost of my time ...

So, I bought a Sparkfun Esp32 Thing. It works great. TFT_Espi and lvgl started up pretty quickly and my Mandelbrot code ran *far* faster than on the Nano.

The bit I'm working on at the moment is getting my display to work on battery power. There's a jumper on the screen to change it from 5v to 3.3v. I'm also looking at how I calculate the percentage remaining charge in the battery. Once I've done those two things, my final bit of hardware is working out how I get round the need to hold down the '0' button every time I want to upload code.

On the software side, I've started a new Platform.IO project in VS Code on my Mac, and am assembling my template - abstracting screen stuff away etc. I don't think I've set up Platform.IO property - it can't see TFT_Espi or lvgl header files yet.

Thanks for your advice.

Ian.

**Comment by Bodmer** [comment]
Good choice.

I like your Mandlebrot sketch on Github, nice colour scheme. I changed the doubles to floats since the ESP32 has an FPU which makes floats much faster and it filled the screen image in 700ms (480x320 screen, 40MHz SPI clock), or 610ms with line by line DMA. Since the colour resolution and iteration count is low (at 60) I suspect the double FP calcs don't improve the image. Very nice sketch.

**Reply by Bodmer** [reply]
By making a few more tweaks and making use of symmetry about the y axis to halve the calculation count, full screen render is down to 279ms which make a "realtime" zoom along the y=0 line practical. Thanks for the sketch it's been fun running it.

**Reply by Bodmer** [reply]
Had a go at a slow zoom and you get some very pretty results with your colour palette:
![screenshot_2021_12_12_18_24_15](https://user-images.githubusercontent.com/15824805/145730736-8c355447-3399-45ba-9f0f-6f3974dc490d.png)

Trying different colour palletes gives some nice results too.
![screenshot_2021_12_12_21_38_4](https://user-images.githubusercontent.com/15824805/145730981-b95fbbf3-db26-4b4a-b6d3-fccbbb764e76.png)

**Comment by IanBUK** [comment]
The set is something I’ve tried in most languages/platforms I’ve worked in. Well, apart from Oracle and SqlServer.  

I had a go in JavaScript, that turned in WebGL fragment shaders. It allowed pinch zoom and panning on a touchscreen, with minimal lag. This was in 2014 on a then current i7 laptop. I had a second iframe, which used the Babylon library to build a 3D mesh, allowing you to spin/ zoom and add different renderers. 

All good fun.
